src/ripple/app/ledger/impl/InboundLedgers.cpp

```diff
@@ -69,23 +69,51 @@ class InboundLedgersImp : public InboundLedgers
         std::uint32_t seq,
         InboundLedger::Reason reason) override
     {
+        std::stringstream ss;
+        ss << "InboundLedger::acquire: "
+           << "Request: " << to_string(hash) << ", " << seq
+           << " NeedNetworkLedger: "
+           << (app_.getOPs().isNeedNetworkLedger() ? "yes" : "no")
+           << " Reason: " << to_string(reason) << " Old rule: ";
+        if (app_.getOPs().isNeedNetworkLedger() &&
+            (reason != InboundLedger::Reason::GENERIC) &&
+            (reason != InboundLedger::Reason::CONSENSUS))
+            ss << "false";
+        else
+            ss << "true";
+
         assert(hash.isNonZero());
         assert(
             reason != InboundLedger::Reason::SHARD ||
             (seq != 0 && app_.getShardStore()));
 
-        // probably not the right rule
-        if (app_.getOPs().isNeedNetworkLedger() &&
-            (reason != InboundLedger::Reason::GENERIC) &&
-            (reason != InboundLedger::Reason::CONSENSUS))
-            return {};
+        bool const isFull = app_.getOPs().isFull();
+        bool const fallingBehind = app_.getOPs().isFallingBehind();
+        LedgerIndex const validSeq =
+            app_.getLedgerMaster().getValidLedgerIndex();
+        constexpr std::size_t lagLeeway = 20;
+        bool const nearFuture =
+            (seq > validSeq) && (seq < validSeq + lagLeeway);
+        bool const consensus = reason == InboundLedger::Reason::CONSENSUS;
+        bool const shouldAcquire =
```

Comment:
I think this logic is OK, but I suggest putting comments and/or cleaning up a bit as follows:

Namely, nearFuture is if:
Requested seq > valid seq BUT requested seq is < 20 greater than the valid seq. It means the requested seq is in the near future relative to the validated ledger. The requested is  is not behind or even, but between 1 and 19 inclusive ledgers ahead of the valid ledger. Therefore, I shouldn't be trying to acquire this ledger. I'm not there yet, and it's possible/likely I have the tx's necessary to build it if I ever need to, and it might not even become validated. On the other hand, if it's 20 or so in the future, I should request it so that I can jump ahead because it probably means I'll have a hard time catching up otherwise.

fallingBehind is if the last closed ledger is at least 2 behind the validated ledger. I am falling behind the network. The reason this should not simply be only at least 1 behind the validated ledger is that I think a slight lag is a normal case because some nodes get there slightly later than others. But a difference of 2 means that at least a full ledger interval has passed, so we are beginning to fall behind.

We shouldn't be acquiring this ledger if we're full. Assume we are functioning correctly and have the necessary tx's for when the time comes.

So, here's what I think the line is checking for. Namely, acquire the ledger if any of the following are false:
    1) We're in sync, or
    2) we are not falling behind, or
    3) the request is for a ledger in the near future, or
    4) it's a Reason::CONSENSUS request.
    
Please confirm the reason for checking the ledger request type Reason::CONSENSUS.
As written, I think it means to always acquire an inbound ledger if it's for something other than CONSENSUS. In other words, only suppress CONSENSUS ledgers (if other conditions are met).

Since this is all boolean math, it easily could be the opposite of everything I just wrote if 1 thing is wrong.

It might make sense to re-write this line to something much easier to reason about, for future generations, after comments like I've described.

Perhaps something like a lambda with this logic:

if not full:
    return true
if falling behind:
    return true
if request for ledger NOT in near future:
  return true
if NOT a Reason::CONSENSUS ledger:
  return true

return false

---------------------

Reply:
The logic you outlined isn't quite right, but I pushed a commit to update and add comments to explain all these things. Take a look.